### PR TITLE
Sampling infrastructure + PSF convolution + 3D intensity model

### PR DESCRIPTION
**Pull request by sweverett**
## Summary

Major feature branch adding sampling infrastructure, PSF convolution, and 3D intensity modeling. **Do NOT merge until PR #17 (`tng_setup`) merges and this branch is rebased onto updated main.**

> **Note:** This PR supersedes the `se/sampling` branch (no separate PR needed) — all sampling commits are fully contained here.

### What's new
- **Sampling infrastructure** — `kl_pipe/sampling/` with NumPyro (recommended), emcee, nautilus, BlackJAX backends; `InferenceTask` + `PriorDict` + config system + YAML support
- **PSF convolution** — `kl_pipe/psf.py` with FFT pipeline, oversampled rendering (N=5 default), `PSFData` pytree
- **3D intensity model** — `InclinedExponentialModel` upgraded from 2D thin-disk to 3D sech² vertical profile with k-space FFT rendering matching GalSim's `SBInclinedExponential`
- **Axis convention bugfix** — GalSim vs KL Pipe coordinate convention alignment
- **Likelihood quadrature interpolation** — slice peak finding for high-SNR regimes
- **Pre-commit hook** — auto-installs black formatter + diagnostics pruning

### 13 commits
1. TNG50 data vector generation, augmentation, diagnostics
2. TNG diagnostic plot fixes
3. 3D particle transformation fixes
4. Angular scaling fix for TNG datavectors
5. Updated env/conda-lock + diagnostics collation script
6. Formatter pass
7. Sampling infrastructure (major)
8. TNG sampling tutorial fix
9. PSF convolution (first impl)
10. Black formatting workflow rework
11. Axis convention bugfix + 3D sersic + likelihood quadrature
12. Fix 26 test regressions from 3D model + k-space rewrite
13. Prune-diagnostics makefile target + pre-commit hook

## Review focus

**API:**
- `kl_pipe/psf.py` — PSF convolution API & FFT pipeline
- `kl_pipe/model.py` — `render_image` + `configure_psf` API
- `kl_pipe/sampling/task.py` — `InferenceTask` interface
- `kl_pipe/priors.py` — `PriorDict` API
- ⚠ `kl_pipe/intensity.py` — 3D model (`_render_kspace`, `__call__` quadrature)
- ⚠ `kl_pipe/transformation.py` — axis convention changes
- ⚠ `kl_pipe/likelihood.py` — PSF integration, quadrature interpolation

**Internals:**
- `kl_pipe/sampling/{emcee,nautilus,blackjax,numpyro}.py` — sampler backends
- `kl_pipe/sampling/configs.py` — config dataclasses

**Tests:**
- `tests/test_utils.py` — `TestConfig` tolerance system

## Reviewing diagnostic outputs

After cloning and checking out this branch, run the tests to generate diagnostic plots, then view them in a browser:

```bash
make test            # fast tests (generates likelihood slices, optimizer recovery, PSF regression plots)
make test-sampling   # MCMC diagnostics (corner plots, traces, parameter recovery)
make diagnostics     # collates all plots into an HTML report and opens in browser
```

### Key plots to inspect

| Directory | What to look for |
|-----------|-----------------|
| `tests/out/likelihood_slices/*_psf_*` | Likelihood peaks land on true values for PSF-convolved models |
| `tests/out/likelihood_slices/joint_*` | Joint model slices — all params peak correctly |
| `tests/out/optimizer_recovery/opt_*_psf_*` | Optimizer recovers true params through PSF convolution |
| `tests/out/psf/galsim_regression_oversample_*` | Oversampled rendering matches GalSim to <5e-4 |
| `tests/out/psf/oversample_convergence_*` | Error decreases monotonically with oversample factor |
| `tests/out/sampling/joint_*` | Corner plots + parameter recovery panels for joint model MCMC |
| `tests/out/intensity/` | 3D intensity model renders (k-space vs quadrature agreement) |

## Test plan
- [ ] `make test` passes (excl. slow + tng_diagnostics)
- [ ] `make test-basic` passes (no TNG data needed)
- [ ] `make test-sampling` passes
- [ ] Inspect diagnostic plots via `make diagnostics`
- [ ] Rebase onto main after PR #17 merges, then re-run tests

### CODE REVIEW

File: kl_pipe/tng/data_vectors.py [hunk 1]
```diff
@@ -670,10 +670,10 @@ def _apply_new_orientation(
         # Step 3: Apply weak lensing shear
         # ===================================================================
         if g1 != 0 or g2 != 0:
-            # Shear matrix (inverse of cen2source)
+            # source→cen = A^{-1} = norm * [[1+g1, g2], [g2, 1-g1]]
```
Comment (pranjalrs) [inline kl_pipe/tng/data_vectors.py:673]: Norm should have `np.sqrt` to be consistent with the shear transformation definition.

**Review by pranjalrs — APPROVED** [review]
One minor comment about shear transfomations, all other (public facing) changes look fine overall.

File: kl_pipe/model.py [hunk 2]
```diff
@@ -256,39 +389,208 @@ def evaluate_circular_velocity(
...
+        flux_theta_override : jnp.ndarray, optional
+            Intensity params for joint mode flux weighting.
 
         Returns
         -------
         jnp.ndarray
-            2D velocity or speed map with shape image_pars.shape.
+            2D velocity or speed map (coarse-scale).
         """
+        if self._psf_data is not None and self._psf_oversample > 1:
```
Comment (pranjalrs) [inline kl_pipe/model.py:548]: PSF convolution for the velocity map fails when using flux image from TNG (likely due to really large flux values). 
The above is not an issue when generating mock TNG velocity fields probably because tng.data_vectors uses a different routine (a numpy based fft convolution).

File: kl_pipe/sampling/numpyro.py [hunk 3]
```diff
@@ -0,0 +1,600 @@
...
+                # Fallback: compute from z samples
+                z_samples = np.array(samples_dict[f"_z_{name}"]).flatten()
+                loc, scale = self._reparam_scales[name]
+                samples_list.append(loc + scale * z_samples)
+
+        samples = np.column_stack(samples_list)
+
+        # Compute log probabilities for samples
+        log_posterior_fn = self.task.get_log_posterior_fn()
+        log_probs = np.array(
```
Comment (pranjalrs) [inline kl_pipe/sampling/numpyro.py:562]: Use jax.vmap instead of list comprehension for speedup

**Review by JiachuanXu — APPROVED** [review]
